### PR TITLE
new plugin to make a playlist from a textfile out of your collection

### PR DESCRIPTION
**Pull request by jmwatte**
I've been using and polishing this for a few days. If anyone has some improvement tips... welcome.

this reads a textfile of the following formats
        `first line:` contains the title of the playlist
        `second line :` expresses the format of the rest of the list. ex:
-  **x. album:  - albumartist: (x)**  =>ex:   2. White - The Beatles (1967)
-  **albumartist: -- album: [x]**  =>ex:      The Beatles -- white [1967]
-  **artist: ; title:**            =>ex:      John Lennon ; imagine
  
  where `x.` represents the linenumberings if the list has that.
  `(x)` or `[x]` is anything in between brackets at the end of the line
  `artist:`,`albumartist:`,`title:`,`album:` should be obvious
        all the other lines follow the format from this  line

It's handy with all these end-of-year lists to see what you got in your collection.
grtz
jm

### CODE REVIEW

File: beetsplug/txt2pls.py [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+    include = set().union(string.ascii_lowercase, string.digits)
+
+    def commands(self):
+        txt2pls = ui.Subcommand('txt2pls',
+                                help='saves a playlist from a textfile\
+                                passed as argument.')
+        txt2pls.func = self.txt2pls_cmd
+        return [txt2pls]
+
+    def cleanLineNrs(self, sourcelines):
```
Comment (sampsyo) [inline beetsplug/txt2pls.py]: According to PEP8, method names should be `snake_case`.

**Comment by sampsyo** [comment]
Seems awesome! So is the idea that it produces an `.m3u` file? I wonder what other interesting things could done with text-based playlist import.

**Comment by jmwatte** [comment]
Yes, it produces a .m3u file.
 I am trying to make behave it beter, like get rid of typing errors in the sourcelist. And  especially with things like a list of Composer - Piece in classical music. I mean everything is nice when all is organised in a consistent way. :)
Well, I was also wondering if there was a way to put things in flexfields
like  a flexfield
```Best of```
and then 1 item could be ```2015 Rolling Stone``` ```2015 Melodymaker```

So that would be a flexfields with a list of multiple values.
Is that already possible?

**Comment by sampsyo** [comment]
Sadly, attributes can't have multiple values. But maybe a delimiter-separated string would work?

**Comment by jmwatte** [comment]
Just a quick question?
If I got it right...
I can set a flex attribute on an item with `` __setitem__(self, key, value)`` from my code
but I don't see anything like that for an album.
Is that correct? Albums don't have flexatt?
grtz
jm

**Comment by jmwatte** [comment]
Did an experiment.
```
for obj in objs:
            obj.__setitem__(self.tag, value)
            if ui.show_model_changes(obj):
                changed.add(obj)
        if not ui.input_yn('Really modify (Y/n)?'):
            return

        # Apply changes to database and files
        with lib.transaction():
            for obj in changed:
                self._log.debug('saving changes to {}', obj)
                obj.try_sync(ui.should_write())
```

and I changed an album.
```
Modifying 2 objects.
Kendrick Lamar - To Pimp a Butterfly
  bestOf: short
Really modify (Y/n)? y
```

and when I do this, I get the result
```
$ beet ls  -a lamar -f'$bestOf'
short
```

but these things give me nothing in return
```
$ beet ls bestOf:short -a
```
 ```
$ beet ls bestOf:short -a lamar
```
```
$ beet ls bestOf:short -a lamar -f'$bestOf'
```
But of course I wanted the ``beet ls -a bestOf:short `` to work ..

I'm probably doing things I shouldn't, but if someone could shed some light on this, I'd appreciate i.
grts
jmt

**Comment by sampsyo** [comment]
Those queries definitely should work. Does it work to put the `-a` flag after the command?

    $ beet ls -a bestOf:short

**Comment by sampsyo** [comment]
Also, yes, `obj.__setitem__(key, value)` works, but it's longhand for `obj[key] = value`. You probably want to use that. You can read more in the Python documentation for `__setitem__`.

**Comment by jmwatte** [comment]
Strange,
just to be sure. Doing ``item[bestOf] =  'value;'`` or ``album[bestof] = value`` should do the trick right? I can see the fields in the tags, they are there if I use for example the modify command. It lists them all. But I don't get them in the search. I hoped that ``beet ls bestOf:; -a`` would list all albums with ``;`` but I get nothing.
There is nothing else I have to do to be able to search them, right?

**Comment by sampsyo** [comment]
Yep! Either this:

    item['bestOf'] = value

or

    item.bestOf = value

should do it.

You'll probably need to escape `;` as `;` or as in `'bestOf:;'`, since shells want to use it as a separator between commands.

**Comment by jmwatte** [comment]
Very strange. I know the tags are there.
But I can not search for them.
 ```
$ beet ls lamar -a -f'$bestOf'
short;American Songwriter : Top 50 Albums of 2015;Complex: 50 Best Albums of 2015;Consequence of Sound: Top 50 Albums of 2015;crave best of 2015;A.V. Club: 15 Best Albums of 2015;AllMusic: 55 Best Albums of 2015 ;

$ beet ls bestOf:2015 -a

$ beet mod -a bestOf:2015 bestOf=2014
error: No matching albums found.

$ beet mod lamar -a bestOf:2015 bestOf=2014
error: No matching albums found.


$ beet mod lamar -a  bestOf=2014
Modifying 1 albums.
Kendrick Lamar - To Pimp a Butterfly
  bestOf: short;American Songwriter : Top 50 Albums of 2015;Complex: 50 Best Albums of 2015;Consequence of Sound: Top 50 Albums of 2015;crave best of 2015;A.V. Club: 15 Best Albums of 2015;AllMusic: 55 Best Albums of 2015 ; -> 2014
Really modify, move and write tags (Y/n)? 
```

mmmm.....

**Comment by jmwatte** [comment]
Tried it with the modify command on the commandline
```
$ beet mod lamar -a inList=test
Modifying 1 albums.
Kendrick Lamar - To Pimp a Butterfly
  inList: test
Really modify, move and write tags (Y/n)? y

$ beet ls -a inList:test

$ beet ls -a lamar -f'$inList'
test

```
and then tried it on the item level
```
$ beet mod lamar mortal inList=test
Modifying 1 items.
Kendrick Lamar - To Pimp a Butterfly - Mortal Man
  inList: test
Really modify, move and write tags (Y/n)? y
 
$ beet ls inList:test

```

So, as far as I can tell  with flexfields and the modify command the flexfields are there but I cannot use them within a search. If I use ``-f'$flexfield`` I see them.
Is this the intended behaviour?
I'm on the latest from github and later on I'll try to disable all my plugins and try again ...
grtz
jm

**Comment by jmwatte** [comment]
So disabled all plugins and still no search results.

**Comment by sampsyo** [comment]
Thanks for persevering! Incredibly, the bug seems to be related to capitalization of field names:

    $ beet mod full TEST=test
    Modifying 1 items.
    the artist - the album - full
      TEST: test
    Really modify (Y/n)? y
    $ beet ls TEST:test
    $ beet mod full test=test
    Modifying 1 items.
    the artist - the album - full
      test: test
    Really modify (Y/n)? y
    $ beet ls test:test
    the artist - the album - full

Investigating.

**Comment by sampsyo** [comment]
Somewhat unsurprisingly, the problem is that queries are parsed in a case-normalized way (see `key = key.lower()` in `beets.dbcore.queryparse`), but modifications with `beet modify` are not. This means that you were setting a field called `inList` or `itemOf`, but you were (unbeknownst to you) querying the field names `inlist` and `itemof`.

The idea was that you could mess up the case in a query (e.g., `AlbumArtist:beatles`) and still get the behavior you want. This is a nice behavior—IMO, there's not much of a need for fields to be case sensitive (e.g., you don't want `foo` and `Foo` to be *different* fields). So there are a few different things we could do:
* Just make `modify` lower-case fields it sets.
* Change the query infrastructure to *ignore* field case, so `itemof` also works to query a field called `itemOf`.
* If we do the above, we should probably do the same for template formatting.
Any opinions?

**Comment by jmwatte** [comment]
Keep the query-all-to-lower. If people want alBumartist they should get want they do get. I guess that leaves only modify-all-to-lower on the table. That gives us the same behaviour. And of course someday someone is gonna go: ' Hey, why does sunsHopper list the same results as sunShopper? Those are 2 different fields!' But I see the dilemma. .. go for intent of for exactness. Getting older I choose for auto-correct. 
What's your preference, master? (star-wars-voice)

**Comment by sampsyo** [comment]
Well, I'm a little undecided. There's:

* The easy way: All fields are lower-case.
* The (probably) right way: Fields are case-insensitive but case-preserving, so you can call your fields "fooBar" if you want but that doesn't matter.

The difference is pretty small, but one takes a bit more effort than the other.

**Comment by jmwatte** [comment]
the right way is the only way but probably takes more effort,

**Comment by jtpavlock** [comment]
closing as abandoned

**Pull request closed without merging**